### PR TITLE
Sort map fields for TextFormat.

### PR DESCRIPTION
**Pull request by thomasvl**
Since Bool isn't Comparable, add a sorting function to MapKeyType and use that
in TextFormatEncodingVisitor.

### CODE REVIEW

**Comment by thomasvl** [comment]
I'm not sure this is the best way to do this, but this is what occurred to me last night.

I'm not thrilled with it having to make something `public`, but I didn't see a better way to do this that let us restrict the needed definitions to something within the library since the `Visitor` api itself is public.

Fixes #1075

File: Sources/SwiftProtobuf/FieldTypes.swift [hunk 1]
```diff
@@ -50,6 +50,17 @@ public protocol FieldType {
 /// Marker protocol for types that can be used as map keys
 ///
 public protocol MapKeyType: FieldType {
+    /// A comparision function for where order is needed.  Can't use `Comparable`
+    /// because `Bool` doesn't conform, and since it is `public` there is no way
+    /// to add a conformance internal to SwiftProtobuf.
+    static func lessThan(lhs: BaseType, rhs: BaseType) -> Bool
```
Comment (allevato) [inline Sources/SwiftProtobuf/FieldTypes.swift:56]: We could make the name of the method `_lessThan` to make it clear that it's only meant to be used by the library.

(We probably should have done that for the protocol itself, in retrospect.)

File: Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift [hunk 2]
```diff
@@ -611,9 +611,10 @@ internal struct TextFormatEncodingVisitor: Visitor {
   private mutating func _visitMap<K, V>(
     map: Dictionary<K, V>,
     fieldNumber: Int,
+    lessThanFunc: (K, K) -> Bool,
```
Comment (allevato) [inline Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:614]: Nit: `isOrderedBefore` would be a cleaner name that matches the argument name used by Swift's stdlib for other sorting predicates.

**Review by allevato — APPROVED** [review]

File: Sources/SwiftProtobuf/FieldTypes.swift (hunk 1 above)
Comment (thomasvl) [reply Sources/SwiftProtobuf/FieldTypes.swift:56]: Some of these actually leak into `Visitor` interface, so they sorta have to be `public` to allow someone to write a custom `Vistor`.

Prefixed this one.

File: Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift (hunk 2 above)
Comment (thomasvl) [reply Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:614]: done.

**Review by tbkka — APPROVED** [review]
Looks good.

**Comment by thomasvl** [comment]
Talking with @allevato in person, we realized we might be able to do this like the Custom JSON support with a protocol internally to the library.  I'll take a shot at that before moving ahead with this since it would avoid adding to the public interfaces.

**Comment by thomasvl** [comment]
And that isn't gonna work because I'd need an associated type, so I guess we're sorta stuck exposing something else at the moment. :(

**Comment by thomasvl** [comment]
Landing since no one has come up with a better idea.  We can always revisit if something better comes along (especially if we haven't cut a release), but hopefully the underscore symbols gives up a little wiggle room.

**Pull request merged**